### PR TITLE
Add external access code action operation for renaming after code actions

### PR DESCRIPTION
**Pull request by PranavSenthilnathan**
Roslyn will invoke navigation or inline renaming after code actions if a resultant document from the code action has a navigation token or rename token in the syntax tree. For languages that don't use syntax trees (TypeScript), Roslyn only supports navigation by position. This PR adds inline renaming by position for these languages.

The new external access operation is similar to the [`DocumentNavigationOperation`](https://github.com/dotnet/roslyn/blob/5faf7b5a3bb6a02ac22f80fa00fb9c8ba2ac153e/src/Features/Core/Portable/Common/DocumentNavigationOperation.cs) which is special cased to navigate to a position after a code action.

### CODE REVIEW

File: src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptInlineRenameOperation.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    /// When <see cref="CodeAction.GetOperationsAsync(System.Threading.CancellationToken)"/> is called an implementation
+    /// of <see cref="CodeAction"/> can return an instance of this operation along with the other 
+    /// operations they want to apply. For example, an implementation could generate a new <see cref="Document"/>
+    /// in one <see cref="CodeActionOperation"/> and then have the host editor navigate to that
+    /// <see cref="Document"/> and invoke rename at a given position using this operation.
+    /// </summary>
+#pragma warning restore RS0030 // Do not used banned APIs
+    internal class VSTypeScriptInlineRenameOperation(DocumentId documentId, int position)
+        : DocumentNavigationOperation(documentId, position)
+    {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptInlineRenameOperation.cs:21]: i would not do it this way.  i would just make a whole new operation just for the purpose of invoking the inline rename system.

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs [hunk 2]
```diff
@@ -301,6 +302,11 @@ internal class CodeActionEditHandlerService(
                 var navigationService = workspace.Services.GetRequiredService<IDocumentNavigationService>();
                 await navigationService.TryNavigateToPositionAsync(
                     this._threadingContext, workspace, navigationOperation.DocumentId, navigationOperation.Position, cancellationToken).ConfigureAwait(false);
+                if (navigationOperation is VSTypeScriptInlineRenameOperation)
+                {
+                    var openDocument = workspace.CurrentSolution.GetRequiredDocument(navigationOperation.DocumentId);
+                    _renameService.StartInlineSession(openDocument, new TextSpan(navigationOperation.Position, 0), cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:308]: this would move into the Apply method of the new operation.

**Review by sharwell — DISMISSED** [review]
`CodeActionOperation` is already a public API which allows `Apply` to be overridden. It should not be necessary to create a new external access API when the public API already exists.

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs (hunk 2 above)
Comment (PranavSenthilnathan) [reply src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:308]: I tried this approach but having the operation execute with all the other operations causes problems. Specifically, it looks like the undo transaction open while running the operation gives an exception: 
`System.Runtime.InteropServices.COMException: 'Catastrophic failure (Exception from HRESULT: 0x8000FFFF (E_UNEXPECTED))'`

which is coming from:
`Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.InlineRename.VisualStudioInlineRenameUndoManagerServiceFactory.InlineRenameUndoManager.GetUndoUnits(Microsoft.VisualStudio.OLE.Interop.IOleUndoManager undoManager)`

When I remove the transaction (actually just moved the commit before all the operations run), there is no error. So it seems like there's not a clean way of doing this without making at least some change to the CodeActionEditHandlerService.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:308]: interesting.  and that's a fair point about a problem with the current system.  it presumes the apply calls will just make edits that can then be undone in the global undo transaction.

So an alternative i'd be ok with would be to make an internal, sealed `StartInlineRenameSessionOperation` class that is a sibling to DocumentNavigationOperation.  

we'd special case looking for that and starting the inline rename session.  You could then expose that operation in your EA layer with some sort of factory-method.

Comment (sharwell) [reply src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs]: @PranavSenthilnathan Can you link to the code where the transaction is created and completed? I'm specifically wondering what change you made here:
> When I remove the transaction (actually just moved the commit before all the operations run)

**Comment by PranavSenthilnathan** [comment]
> `CodeActionOperation` is already a public API which allows `Apply` to be overridden. It should not be necessary to create a new external access API when the public API already exists.

The problem I was running into with the undo manager: https://github.com/dotnet/roslyn/pull/70531#discussion_r1370539603. As Cyrus mentioned, I think this comes down to the issue that this code action operation is not just a pure edit and messes with the undo manager.

File: src/Features/Core/Portable/Common/StartInlineRenameSessionOperation.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    /// of <see cref="CodeAction"/> can return an instance of this operation along with the other 
+    /// operations they want to apply. For example, an implementation could generate a new <see cref="Document"/>
+    /// in one <see cref="CodeActionOperation"/> and then have the host editor navigate to that
+    /// <see cref="Document"/> and invoke rename at a given position using this operation.
+    /// </summary>
+#pragma warning restore RS0030 // Do not used banned APIs
+    internal class StartInlineRenameSessionOperation(DocumentId documentId, int position) : CodeActionOperation
+    {
+        internal DocumentId DocumentId { get; } = documentId ?? throw new ArgumentNullException(nameof(documentId));
+        internal int Position { get; } = position;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Common/StartInlineRenameSessionOperation.cs:23]:
```suggestion
        public DocumentId DocumentId { get; } = documentId ?? throw new ArgumentNullException(nameof(documentId));
        public int Position { get; } = position;
```

File: src/Features/Core/Portable/Common/StartInlineRenameSessionOperation.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    /// <summary>
+    /// A <see cref="CodeActionOperation"/> for navigating to a specific position in a document and invoking inline rename.
+    /// When <see cref="CodeAction.GetOperationsAsync(System.Threading.CancellationToken)"/> is called an implementation
+    /// of <see cref="CodeAction"/> can return an instance of this operation along with the other 
+    /// operations they want to apply. For example, an implementation could generate a new <see cref="Document"/>
+    /// in one <see cref="CodeActionOperation"/> and then have the host editor navigate to that
+    /// <see cref="Document"/> and invoke rename at a given position using this operation.
+    /// </summary>
+#pragma warning restore RS0030 // Do not used banned APIs
+    internal class StartInlineRenameSessionOperation(DocumentId documentId, int position) : CodeActionOperation
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Common/StartInlineRenameSessionOperation.cs:20]:
```suggestion
    internal sealed class StartInlineRenameSessionOperation(DocumentId documentId, int position) : CodeActionOperation
```

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptRenameOperationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.CodeActions;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api
+{
+    internal static class VSTypeScriptRenameOperationFactory
+    {
+        public static CodeActionOperation CreateRenameOperation(DocumentId documentId, int position) => new StartInlineRenameSessionOperation(documentId, position);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/VSTypeScriptRenameOperationProvider.cs:11]:
```suggestion
        public static CodeActionOperation CreateRenameOperation(DocumentId documentId, int position)
            => new StartInlineRenameSessionOperation(documentId, position);
```

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs (hunk 2 above)
Comment (PranavSenthilnathan) [reply src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs]: It was moving https://github.com/dotnet/roslyn/blob/afabc2f93fb807375f6fdb3b8ee9dfdd0ac6c1dd/src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs#L179 right above the `try` which is a non-sensical change but I just wanted to see if the transaction was causing the issue.

**Comment by PranavSenthilnathan** [comment]
I think all the comments have been addressed so this should be ready for merge.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by PranavSenthilnathan** [comment]
After merging main and retriggering CI, completely different tests are failing than before merging :upside_down_face:

**Pull request merged**